client.go

```diff
@@ -85,6 +91,9 @@ func WithVersion(version string) ClientOption {
 }
 
 func (c *Client) request(ctx context.Context, method string, urlStr string, queryParams map[string]string, requestBody interface{}) (*http.Response, error) {
+	c.requestMutex.Lock()
+	defer c.requestMutex.Unlock()
```

Comment:
What variables exactly are subject to a race condition?

Reply:
I'm concerned that if multiple goroutines are concurrently using the client to make requests and the Notion api starts rate limiting, one goroutine might always be faster than the others and end up being always served while the other goroutines starve (and consequently fail after retrying a few times). What I want is for only one goroutine to be waiting for retry at a time, rather than all of them waiting roughly the same amount of time and retrying all at once. I'm not sure if the scenario I'm imagining is possible, but hopefully that makes sense.

Reply:
> one goroutine might always be faster than the others and end up being always served while the other goroutines starve (and consequently fail after retrying a few times)

When you say that one goroutine is always faster, I guess you mean that it's sending requests to the Notion API more frequently? If so, then this locking wouldn't prevent it from getting another lock faster then another goroutine. The other goroutine in that case won't be able to send a request at all, whereas without locking it would be able to at least try.

Reply:
I don't mean sending requests to the Notion API more frequently. I mean if two goroutines try to make a request at nearly the same time and both get rate limited, the one that was slightly faster will retry first and (probably) succeed, then the other will retry almost immediately after and may get rate limited again. Now if there are four goroutines all making requests at nearly the same time, the one that is the slowest will fail due to getting rate limited 3 times. 

Also, if one goroutine sends requests more frequently, I don't think locking will prevent the other goroutine from sending requests completely. Here's an example: https://go.dev/play/p/RosOP4dzTfY
I might be misunderstanding though.

Reply:
Thanks for the explanation!

As I'm understanding it, we're basically trying to coordinate the requests. However, IMHO, that's not an API client's responsibility. Someone may also multiple clients, which would then void the effect of request locking.

Reply:
Ok I see, that makes sense